velox/expression/CastExpr.cpp

```diff
@@ -520,63 +520,73 @@ VectorPtr CastExpr::applyDecimal(
   return castResult;
 }
 
-/// Apply casting between a custom type and another type.
-/// @param castTo The boolean indicating whether to cast an input to the custom
-/// type or from the custom type
-/// @param input The input vector
-/// @param allRows The selectivity vector of all rows in input
-/// @param nonNullRows The selectivity vector of non-null rows in input
-/// @param castOperator The cast operator for the custom type
-/// @param thisType The custom type
-/// @param otherType The other type involved in this casting
-/// @param context The context
-/// @param nullOnFailure Whether this is a cast or try_cast operation
-/// @param result The output vector
-template <bool castTo>
-void applyCustomTypeCast(
-    VectorPtr& input,
-    DecodedVector& inputDecoded,
-    const SelectivityVector& allRows,
-    const SelectivityVector& nonNullRows,
-    const CastOperatorPtr& castOperator,
-    const TypePtr& thisType,
-    const TypePtr& otherType,
+void CastExpr::applyComplexType(
+    const SelectivityVector& rows,
+    const BaseVector& input,
     exec::EvalCtx& context,
-    bool nullOnFailure,
+    const TypePtr& fromType,
+    const TypePtr& toType,
     VectorPtr& result) {
-  VELOX_CHECK_NE(
-      thisType,
-      otherType,
-      "Attempting to cast from {} to itself.",
-      thisType->toString());
-
-  exec::LocalSelectivityVector baseRows(
-      *context.execCtx(), inputDecoded.base()->size());
-  baseRows->clearAll();
-  context.applyToSelectedNoThrow(nonNullRows, [&](auto row) {
-    baseRows->setValid(inputDecoded.index(row), true);
-  });
-  baseRows->updateBounds();
+  switch (toType->kind()) {
+    // Handle complex type conversions
+    case TypeKind::MAP:
+      result = applyMap(
+          rows,
+          input.asUnchecked<MapVector>(),
+          context,
+          fromType->asMap(),
+          toType->asMap());
+      break;
+    case TypeKind::ARRAY:
+      result = applyArray(
+          rows,
+          input.asUnchecked<ArrayVector>(),
+          context,
+          fromType->asArray(),
+          toType->asArray());
+      break;
+    case TypeKind::ROW:
+      result = applyRow(
+          rows,
+          input.asUnchecked<RowVector>(),
+          context,
+          fromType->asRow(),
+          toType->asRow());
+      break;
+    case TypeKind::SHORT_DECIMAL:
```

Comment:
I agree. Decimal needs to take the same path with complex types and custom types (as opposed to other primitive types) because they all peel off encodings before processing the data, and thus I combined these paths to reduce code duplicates. What would you suggest as the name of a function that handles Decimal, complex, and custom types? (If we don't want to put them into one function, there will be some code duplicates.)

Reply:
I wonder why primitive types (other than decimal) are handled differently. I expect all types need to be processed the same way using the workload I mentioned earlier. Am I missing something?

Reply:
I'm not sure why either because those are existing code. I guess it's because peeling off encoding improves the performance of processing complex-typed data more significantly, so it was worth doing.

Reply:
Another question is whether expression evaluation should peel off encodings and exclude nulls before calling evalSpecialForm. If it did that, then special forms like CAST would not need to duplicate that logic.

CC: @bikramSingh91

Reply:
To answer this question, I first tried to look as to why we need special form expr in the first place. Its because they don’t follow the regular flow of expr eval as defined in expr.cc and we can arbitrarily handle inputs as they wish.
This works perfectly for others like conjuncts or coalesce but cast seems to be the only one which has a flow very similar to the regular flow but with some added steps like extracting internal vectors from complex types. I believe adding a mechanism to allow special forms to reuse the flow might end up with a solution over-fitted to cast only, so might be an overkill unless we have more use cases like this. However, I feel there is definitely value in modularizing functionality like peeling and null handling in evalEncodings and evalWithNulls respectively that we can re-use here. The functionality in evalEncodings is also repeated somewhat in applyFunctionWithPeeling and hence can also be reused there. Moreover, like you suggested earlier, we can test those modular functions separately to make sure their functionality is bullet proof